cmd/clusterctl/client/repository/repository_github.go

```diff
@@ -284,23 +319,41 @@ func (g *gitHubRepository) downloadFilesFromRelease(release *github.RepositoryRe
 		return nil, errors.Errorf("failed to get file %q from %q release", fileName, *release.TagName)
 	}
 
-	reader, redirect, err := client.Repositories.DownloadReleaseAsset(context.TODO(), g.owner, g.repository, *assetID, http.DefaultClient)
-	if err != nil {
-		return nil, g.handleGithubErr(err, "failed to download file %q from %q release", *release.TagName, fileName)
-	}
-	if redirect != "" {
-		req, err := http.NewRequestWithContext(ctx, http.MethodGet, redirect, http.NoBody)
-		if err != nil {
-			return nil, errors.Wrapf(err, "failed to download file %q from %q release via redirect location %q: failed to create request", *release.TagName, fileName, redirect)
+	var reader io.ReadCloser
+	var retryError error
+	_ = wait.PollImmediate(retryableOperationInterval, retryableOperationTimeout, func() (bool, error) {
+		var redirect string
+		var downloadReleaseError error
+		reader, redirect, downloadReleaseError = client.Repositories.DownloadReleaseAsset(context.TODO(), g.owner, g.repository, *assetID, http.DefaultClient)
+		if downloadReleaseError != nil {
+			retryError = g.handleGithubErr(downloadReleaseError, "failed to download file %q from %q release", *release.TagName, fileName)
+			// return immediately if we are rate limited
+			if _, ok := downloadReleaseError.(*github.RateLimitError); ok {
+				return false, retryError
+			}
+			return false, nil
 		}
-
-		response, err := http.DefaultClient.Do(req) //nolint:bodyclose // (NB: The reader is actually closed in a defer)
-		if err != nil {
-			return nil, errors.Wrapf(err, "failed to download file %q from %q release via redirect location %q", *release.TagName, fileName, redirect)
+		if redirect != "" {
+			req, err := http.NewRequestWithContext(ctx, http.MethodGet, redirect, http.NoBody)
+			if err != nil {
+				retryError = errors.Wrapf(err, "failed to download file %q from %q release via redirect location %q: failed to create request", *release.TagName, fileName, redirect)
+				return false, nil
+			}
+
+			response, err := http.DefaultClient.Do(req) //nolint:bodyclose // (NB: The reader is actually closed in a defer)
+			if err != nil {
+				retryError = errors.Wrapf(err, "failed to download file %q from %q release via redirect location %q", *release.TagName, fileName, redirect)
+				return false, nil
+			}
+			reader = response.Body
 		}
-		reader = response.Body
-	}
+		retryError = nil
+		return true, nil
+	})
 	defer reader.Close()
+	if retryError != nil {
+		return nil, retryError
+	}
```

Comment:
```suggestion
	if retryError != nil {
		return nil, retryError
	}
	defer reader.Close()
```

Shouldn't the defer block be after the error handling? If we receive a retryError it means that the `reader` is most likely nil. If the reader is nil then having this defer block above the error handling will cause a panic.

Reply:
Because the `reader` reference is established outside of the retry func, the easiest thing to to is probably check for nil before deferred closing. We could also do the defer + close in each retry func loop, but we'd still have to maintain an outer variable and assign the return value from the download/redirect func to it, and we'd still need to check for nil before closing that outer `io.ReadCloser` object. So I think the easiest approach is just to check for not nil before invoking `.Close()`